cmd/root.go

```diff
@@ -0,0 +1,71 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/tlkamp/plugin-cli-example/cmd/plugins"
+	"github.com/tlkamp/plugin-cli-example/cmd/version"
+	"github.com/tlkamp/plugin-cli-example/internal/executor"
+)
+
+type PluginExecutor interface {
+	// Lookup returns true if the plugin is found by the plugin executor
+	Lookup(name string) bool
+
+	// Execute executes the plugin by the given name
+	Execute(name string, args []string, env []string) error
+}
+
+func Execute() error {
+	var (
+		pluginDIR    string
+		pluginPrefix string
+	)
+
+	rootCmd := &cobra.Command{
+		Use:   "cli-with-plugins",
+		Short: "A demo of how to create a CLI that supports arbitrary plugins",
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return fmt.Errorf("error getting current user: %w", err)
+	}
+
+	dd := strings.Join([]string{u.HomeDir, ".plugin-cli-example/plugins"}, string(os.PathSeparator))
+
+	rootCmd.PersistentFlags().StringVar(&pluginDIR, "plugin-dir", dd, "the plugin directory to use")
+	rootCmd.PersistentFlags().StringVar(&pluginPrefix, "plugin-prefix", "example", "the plugin prefix to use")
```

Comment:
discussion: I'm not sure I like allowing the user to specify an arbitrary plugin prefix. I think it could result in weird things where the CLI attempts to execute something that isn't designed to be called as a plugin.